File: 349_intersection_of_2_arrays.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;


class Solution {
public:
    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res;
        unordered_set<int> ref;
        for(const int& n:nums1) {
            ref.insert(n);
        }
        for(const int& n:nums2) {
            auto found = ref.find(n);
            if(found != ref.end()){
                res.push_back(n);
                ref.erase(found);
            }
        }
        return res;
    }
};

int main() {
    vector<int> nums1 = {1, 2, 2, 1};
    vector<int> nums2 = {2, 2};
    Solution sol;
    vector<int> intersection = sol.intersection(nums1, nums2);
    for(const int& n:intersection) {
        cout << n << " ";
    }
    cout << endl;
    return 0;
}

File: 349_intersection_of_2_arrays.py
class Solution(object):
    def intersection(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        return list(set(nums1)&set(nums2))
#        if len(nums1)>len(nums2):
#            numset = set(nums2)
#            numarr = nums1
#        else:
#            numset = set(nums1)
#            numarr = nums2
#        res = []
#        for n in numarr:
#            if n in numset:
#                res.append(n)
#                numset.remove(n)
#        return res

sol = Solution()
nums1 = [1,2,2,1]
nums2 = [2,2]
assert(sol.intersection(nums1,nums2)==[2])

